### PR TITLE
feat: submission callback

### PR DESCRIPTION
**Pull request by dextertanyj**

### CODE REVIEW

File: backend/src/judge/judge.service.ts [hunk 1]
```diff
@@ -42,76 +81,164 @@ export class JudgeService {
...
+    roomId: string;
+    submissionId: string;
+  } | null> {
+    const submissionId = token;
+
+    const { data } = await this.axiosInstance.get<Judge0Callback>(
+      `/submissions/${submissionId}`,
+    );
+
+    const content = data;
```
Comment (ypinhsuan) [inline backend/src/judge/judge.service.ts:173]:
```suggestion
```

File: backend/src/judge/judge.controller.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  async callback(
+    @Body() response: Pick<Judge0Callback, "token">,
+  ): Promise<void> {
+    const completed = await this.judgeService.handleCallback(response);
+    if (!completed) {
+      return;
+    }
+    const { roomId, submissionId } = completed;
+    this.roomGateway.handleSubmissionUpdate(roomId, submissionId);
+    return;
```
Comment (ypinhsuan) [inline backend/src/judge/judge.controller.ts:29]:
```suggestion
```

File: frontend/src/pages/RoomPage.tsx [hunk 3]
```diff
@@ -213,6 +216,27 @@ export const RoomPage = () => {
...
+      ({ reason }: SubmissionRejectedPayload) => {
+        enqueueSnackbar(`${reason}`, {
+          variant: "error",
+        });
+        setIsSubmitLoading(false);
+      },
+    );
+
+    roomSocket.on(ROOM_EVENTS.SUBMISSION_ACCEPTED, () => {
+      enqueueSnackbar("Processing submission");
```
Comment (ypinhsuan) [inline frontend/src/pages/RoomPage.tsx:230]: Should we make the other user's submit button loading too?
```suggestion
      enqueueSnackbar("Processing submission");
      setIsSubmitLoading(true);
```

File: frontend/src/pages/RoomPage.tsx [hunk 4]
```diff
@@ -213,6 +216,27 @@ export const RoomPage = () => {
...
+    });
+
+    roomSocket.on(
+      ROOM_EVENTS.SUBMISSION_UPDATED,
+      (_payload: SubmissionUpdatedPayload) => {
+        enqueueSnackbar("Submission updated");
+      },
```
Comment (ypinhsuan) [inline frontend/src/pages/RoomPage.tsx:237]:
```suggestion
      },
      setIsSubmitLoading(false);
```

**Review by ypinhsuan — APPROVED** [review]
LGTM

File: backend/src/judge/judge.service.ts [hunk 5]
```diff
@@ -42,76 +81,164 @@ export class JudgeService {
...
     private readonly configService: ConfigService,
     private readonly questionService: QuestionService,
     private readonly redisService: RedisService,
+    private readonly prismaService: PrismaService,
   ) {
     this.axiosInstance = axios.create({
       baseURL: `https://${this.configService.get("judge0.apiHost")}/`,
-      // TODO: Use "wait: false" and provide callback url
-      params: { wait: true, base64_encoded: true, fields: "*" },
+      params: { base64_encoded: true, fields: "*" },
```
[the remark this answers is not in the document]
Comment (dextertanyj) [reply backend/src/judge/judge.service.ts:96]: It is still needed for the `GET` request to retrieve all the other fields, so we can leave it there.

**Review by ypinhsuan — APPROVED** [review]
LGTM

**Pull request merged**